src/symbols_comp.c

```diff
@@ -8,65 +40,94 @@ comp_symbol_list *
 comp_syms_init()
 {
     comp_symbol_list *new_list = malloc(sizeof(comp_symbol_list));
-    new_list->data_count = 0;
-    new_list->data = NULL;
+    tommy_hashtable_init(new_list, HASHTABLE_MAX_SZ);
     return new_list;
 }
 
 void
 comp_syms_clean(comp_symbol_list *list)
 {
-    free(list->data);
+    tommy_hashtable_done(list);
     free(list);
 }
 
 void
 comp_syms_clean_deep(comp_symbol_list *list)
 {
-    for (size_t i = 0; i < list->data_count; ++i)
-    {
-        free(list->data[i]);
-    }
+    tommy_hashtable_foreach(list, comp_syms_clean_one_entry);
     comp_syms_clean(list);
 }
 
 void
 comp_syms_insert(comp_symbol *to_insert, comp_symbol_list *list)
 {
-    size_t curr_count = list->data_count;
-    list->data = realloc(list->data, (curr_count + 1) * sizeof(comp_symbol *));
-    if (list->data == NULL)
-    {
-        err(1, "Error inserting symbol %s in comp_list!",
-            to_insert->sym_ref->sym_name);
-    }
-    list->data[curr_count] = to_insert;
-    list->data_count += 1;
+    tommy_hashtable_insert(list, &to_insert->node, to_insert,
+            hashtable_hash(to_insert->sym_ref->sym_name));
 }
 
 comp_symbol *
-comp_syms_search(const char *to_find, const comp_symbol_list *list)
+comp_syms_search(const char *to_find, comp_symbol_list *list)
 {
-    for (size_t i = 0; i < list->data_count; ++i)
+    comp_symbol* found = tommy_hashtable_search(list, comp_syms_compare,
+            to_find, hashtable_hash(to_find));
+    if (!found)
     {
-        if (!strcmp(list->data[i]->sym_ref->sym_name, to_find))
-        {
-            return list->data[i];
-        }
+        errx(1, "Did not find symbol %s!\n", to_find);
     }
-    errx(1, "Did not find symbol %s!\n", to_find);
+    return found;
 }
 
-comp_symbol_list *
-comp_syms_find_all(const char *to_find, const comp_symbol_list *list)
+comp_symbol **
+comp_syms_find_all(const char *to_find, comp_symbol_list *list)
 {
-    comp_symbol_list *res = comp_syms_init();
-    for (size_t i = 0; i < list->data_count; ++i)
+    comp_symbol** res = calloc(1024, sizeof(comp_symbol*));
+    unsigned int res_sz = 0;
+    tommy_hashtable_node* curr_node = tommy_hashtable_bucket(list, hashtable_hash(to_find));
+    while (curr_node)
     {
-        if (!strcmp(list->data[i]->sym_ref->sym_name, to_find))
+        if (!strcmp(((comp_symbol*) curr_node->data)->sym_ref->sym_name, to_find))
         {
-            comp_syms_insert(list->data[i], res);
+            res[res_sz] = (comp_symbol*) curr_node->data;
+            res_sz += 1;
         }
+        curr_node = curr_node->next;
     }
+    assert(res_sz < 1024);
```

Comment:
Why 1024?

Reply:
A bit of a hack, since earlier I declare a `1024` element size array. I've improved this link in https://github.com/capablevms/CHERI-ELF-comp/pull/39/commits/898e9ecf2a0bc49f012df20ed92b6c49784187a6.

Ideally this would be dynamically increasing, I think, but I decided to use `*_symbol_list **` as return value due to some simplicity. I could potentially use a `tommy_array` here, I think

Reply:
Can we at least make this a named constant so we don't have to change it in multiple places.

Reply:
You mean something beyond the `MAX_FIND_ALL_COUNT` macro I added in https://github.com/capablevms/CHERI-ELF-comp/pull/39/commits/898e9ecf2a0bc49f012df20ed92b6c49784187a6?